### PR TITLE
Graph and show workflow in emacs.

### PR DESCRIPTION
**Pull request by endrebak**
I messed up my git repo, so here I try again. Thanks for the patience. Since emacs and snakemake is all I use, you will probably have to brace yourself for more PRs.

> Hmm, but this is limited to creating the graph for the first rule ofthe Snakefile, right?

Yes, I do not think it is possible to create a graph for anything but the all rule. But if it is, I have never done it and do not know how to do it. I do not yet understand the compile-code either.

> I'd prefer to have this as a customizable variable, as is done with snakemake-program.

I have made `snakemake-graph-program` a defcustom. But I do not know of other pieces of software that can do the same thing though.

### CODE REVIEW

File: snakemake.el [hunk 1]
```diff
@@ -351,6 +356,46 @@ targets."
...
+
+(defun snakemake-show-rulegraph ()
+  "Create and show rulegraph in emacs-buffer."
+  (interactive)
+  (find-file (snakemake-create-rulegraph)))
+
+(defun snakemake-create-dag ()
+  "Create dag for workflow.
+
+  Shows all jobs. If the graph becomes too crowded, use the rulegraph function."
```
Comment (kyleam) [inline snakemake.el]: Elisp docstrings don't indent the second line.  This renders as

    Create dag for workflow.

      Shows all [...]

if in the help buffer.

File: snakemake.el [hunk 2]
```diff
@@ -351,6 +356,46 @@ targets."
...
+  (interactive)
+  (snakemake--create-dotgraph "dag" "png"))
+
+(defun snakemake-create-rulegraph ()
+  "Create rulegraph for workflow.
+
+  Less crowded than dag option."
+  (interactive)
+  (snakemake--create-dotgraph "dag" "png"))
```
Comment (kyleam) [inline snakemake.el]: "dag" rather than "rulegraph" is passed.

File: snakemake.el [hunk 3]
```diff
@@ -351,6 +356,46 @@ targets."
...
+
+(defun snakemake--create-dotgraph (graph-type image-type)
+  "Create a graph for the current snakefile."
+  (let ((dot (shell-command-to-string (concat "which " snakemake-graph-program))))
+    (when (equal dot "")
+      (error (concat "You do not have " snakemake-graph-program " on your emacs path!"))))
```
Comment (kyleam) [inline snakemake.el]: I think a "which" dance like this isn't a typical approach.  If
possible, `call-process` (or some other non-shell-command variant)
is used directly and that signals the error if the program isn't
found.  (There is also `executable-find`.)

File: snakemake.el [hunk 4]
```diff
@@ -351,6 +356,46 @@ targets."
...
+  (let* ((fname (buffer-file-name))
```
Comment (kyleam) [inline snakemake.el]: It'd be nice to support the target at point, like the compile commands
in this file do.

File: snakemake.el [hunk 5]
```diff
@@ -351,6 +356,46 @@ targets."
...
+         (graphname (concat (buffer-file-name) "." image-type))
+         (graph-command (concat "snakemake --" graph-type " --snakefile " fname " | dot -T" image-type " > " graphname)))
+    (progn
+      (shell-command graph-command)
+      ;; hack to return the graph-name from function.
+      (concat graphname))))
```
Comment (kyleam) [inline snakemake.el]: `concat` isn't necessary.

**Comment by kyleam** [comment]
> I messed up my git repo, so here I try again.

OK, thanks.  You can also force push the topic branch rather than
opening up a new PR.

> Since emacs and snakemake is all I use, you will probably have to
> brace yourself for more PRs.

Great, thanks for the PRs and suggestions so far.  And it's nice to
know someone else is using `snakemake{,-mode}.el`.

>> Hmm, but this is limited to creating the graph for the first rule
>> ofthe Snakefile, right?
>
> Yes, I do not think it is possible to create a graph for anything
> but the all rule. But if it is, I have never done it and do not know
> how to do it.

It's specified by the targets passed to Snakemake, in the same way
that you build targets other than the top-most rule.

>> I'd prefer to have this as a customizable variable, as is done with
>> snakemake-program.
>
> I have made snakemake-graph-program a defcustom. But I do not know
> of other pieces of software that can do the same thing though.

Graphviz contains a few other options (see "man dot").  I suppose any
program that understands the dot format as input would work.  But a
defcustom also allows the user to specify a program that isn't in their
`PATH`.

I've tried a slightly different approach in the `km/graph` branch
(d8ddf47).  Thoughts?

**Comment by endrebak** [comment]
Looks good, reading _all_ your code now to get better aquainted with it.

Going to start writing yasnippets for snakemake, do you have any you'd like to share?

**Comment by kyleam** [comment]
> Going to start writing yasnippets for snakemake, do you have any you'd like to share?

That'd be great.  

I only have these [two](https://github.com/kyleam/emacs.d/blob/master/snippets/snakemake-mode/rule) [snippets](https://github.com/kyleam/emacs.d/blob/master/snippets/snakemake-mode/cdrun).  I'm not sure if the second one is generally useful, or even the best way to accomplish that (running a script from it's directory) in Snakemake.

**Pull request closed without merging**